### PR TITLE
upgrade to 0.42

### PR DESCRIPTION
**Pull request by marbar3778**
## Description

Upgrade to protobuf

### CODE REVIEW

File: cmd/ebrelayer/relayer/ethereum.go [hunk 1]
```diff
@@ -364,21 +359,21 @@ func (sub EthereumSub) Replay(fromBlock int64, toBlock int64, cosmosFromBlock in
 	bridgeClaims := sub.getAllClaims(cosmosFromBlock, cosmosToBlock)
 	log.Printf("found out %d bridgeClaims\n", len(bridgeClaims))
 
-	client, err := SetupRPCEthClient(sub.EthProvider)
+	c, err := SetupRPCEthClient(sub.EthProvider)
```
Comment (ElliotFriedman) [inline cmd/ebrelayer/relayer/ethereum.go:362]: This variable has to be renamed. It is so non descriptive. ethRPCClient would work

File: cmd/ebrelayer/relayer/ethereum.go [hunk 2]
```diff
@@ -461,15 +456,15 @@ func (sub EthereumSub) logToEvent(clientChainID *big.Int, contractAddress common
 
 // handleEthereumEvent unpacks an Ethereum event, converts it to a ProphecyClaim, and relays a tx to Cosmos
 func (sub EthereumSub) handleEthereumEvent(events []types.EthereumEvent) error {
-	var prophecyClaims []ethbridge.EthBridgeClaim
+	var prophecyClaims []*ethbridge.EthBridgeClaim
```
Comment (ElliotFriedman) [inline cmd/ebrelayer/relayer/ethereum.go:459]: Why does this array of EthBridgeClaims have to be an array of pointers? cc @iverc @timlind

Comment (ElliotFriedman) [reply cmd/ebrelayer/relayer/ethereum.go:459]: Why can't it just be an array of EthBridgeClaims?

File: cmd/ebrelayer/txs/parser.go [hunk 3]
```diff
@@ -68,14 +68,14 @@ func EthereumEventToEthBridgeClaim(valAddr sdk.ValAddress, event types.EthereumE
 	nonce := int(event.Nonce.Int64())
 
 	// Package the information in a unique EthBridgeClaim
-	witnessClaim.EthereumChainID = chainID
-	witnessClaim.BridgeContractAddress = bridgeContractAddress
-	witnessClaim.Nonce = nonce
-	witnessClaim.TokenContractAddress = tokenContractAddress
+	witnessClaim.EthereumChainId = int64(chainID)
```
Comment (ElliotFriedman) [inline cmd/ebrelayer/txs/parser.go:71]: The old version of this EthBridgeClaim object had the chainid be an int. We should do the same thing to our protos so that we don't have to do this weird casting here.

File: cmd/ebrelayer/txs/parser.go [hunk 4]
```diff
@@ -68,14 +68,14 @@ func EthereumEventToEthBridgeClaim(valAddr sdk.ValAddress, event types.EthereumE
...
+	witnessClaim.BridgeContractAddress = bridgeContractAddress.String()
+	witnessClaim.Nonce = int64(nonce)
```
Comment (ElliotFriedman) [inline cmd/ebrelayer/txs/parser.go:73]: Same thing with nonce. It should just be an int.

Comment (ElliotFriedman) [reply cmd/ebrelayer/txs/parser.go:73]: Did we do this because protobuffs do not support regular ints? cc @timlind 
https://developers.google.com/protocol-buffers/docs/proto3#scalar

File: cmd/ebrelayer/txs/relayToCosmos.go [hunk 5]
```diff
@@ -23,8 +20,8 @@ var (
 
 // RelayToCosmos applies validator's signature to an EthBridgeClaim message containing
 // information about an event on the Ethereum blockchain before relaying to the Bridge
-func RelayToCosmos(cdc *codec.Codec, moniker, password string, claims []types.EthBridgeClaim, cliCtx context.CLIContext,
-	txBldr authtypes.TxBuilder, sugaredLogger *zap.SugaredLogger) error {
+func RelayToCosmos(moniker, password string, claims []*types.EthBridgeClaim, cliCtx client.Context,
```
Comment (ElliotFriedman) [inline cmd/ebrelayer/txs/relayToCosmos.go:23]: Same question here: Why do we need to make it an array of pointers to EthBridgeClaims instead of just an array of EthBridgeClaims?

File: cmd/ebrelayer/types/types.go [hunk 6]
```diff
@@ -40,16 +40,16 @@ func (d Event) String() string {
...
-	ID                    [32]byte
-	From                  common.Address
 	To                    []byte
-	Token                 common.Address
 	Symbol                string
+	EthereumChainID       *big.Int
 	Value                 *big.Int
 	Nonce                 *big.Int
 	ClaimType             ethbridge.ClaimType
+	ID                    [32]byte
```
Comment (ElliotFriedman) [inline cmd/ebrelayer/types/types.go:49]: We should not be changing the order in which items in our structs are declared if we can help it. It seems like this change doesn't really even change the keys or values or their types, instead it just moves things around. We should revert these changes.

File: proto/sifnode/ethbridge/v1/tx.proto [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+// MsgLock defines a message for locking coins and triggering a related event
+message MsgLock {
+  string cosmos_sender     = 1;
+  string amount            = 2 [(gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int",
+    (gogoproto.nullable)   = false,
+    (gogoproto.moretags)   = "yaml:\"amount\""];
+  string symbol            = 3 [(gogoproto.moretags) = "yaml:\"symbol\""];
+  int64  ethereum_chain_id = 4 [(gogoproto.moretags) = "yaml:\"ethereum_chain_id\""];
+  string ethereum_receiver = 5 [(gogoproto.moretags) = "yaml:\"ethereum_receiver\""];
```
Comment (ElliotFriedman) [inline proto/sifnode/ethbridge/v1/tx.proto:27]: Can we use a custom type here like we do with amount, but instead of sdk.int, use ethereum.address? cc @timlind

File: proto/sifnode/oracle/v1/types.proto [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+//TODO: Add parameters to Genesis state ,such as minimum liquidity required to create a pool
+message GenesisState {
+  repeated string address_whitelist = 1;
+  string          admin_address     = 2;
+}
+
+// Claim contrains an arbitrary claim with arbitrary content made by a given validator
+message Claim {
+  string id                = 1;
+  string validator_address = 2;
```
Comment (ElliotFriedman) [inline proto/sifnode/oracle/v1/types.proto:18]: I don't see why we can't use a custom gogo proto type here as well for this instead of just using a string.

File: x/ethbridge/client/cli/tx.go [hunk 9]
```diff
@@ -1,56 +1,57 @@
...
+
 			tokenContract := types.NewEthereumAddress(tokenContractString)
 
 			if !common.IsHexAddress(args[0]) {
 				return errors.Errorf("invalid [bridge-registry-contract]: %s", args[0])
 			}
 			bridgeContract := types.NewEthereumAddress(args[0])
 
-			nonce, err := strconv.Atoi(args[1])
+			nonce, err := strconv.ParseInt(args[1], 10, 64)
```
Comment (ElliotFriedman) [inline x/ethbridge/client/cli/tx.go:54]: Why are we using ParseInt instead of ATOI?

File: x/ethbridge/client/rest/rest.go [hunk 10]
```diff
@@ -93,33 +96,35 @@ func createClaimHandler(cliCtx context.CLIContext) http.HandlerFunc {
...
 			return
 		}
 
+		ct := types.ClaimType(claimType)
+
 		// create the message
 		ethBridgeClaim := types.NewEthBridgeClaim(
-			req.EthereumChainID, bridgeContractAddress, req.Nonce, req.Symbol,
-			tokenContractAddress, ethereumSender, cosmosReceiver, validator, req.Amount, claimType)
+			int64(req.EthereumChainID), bridgeContractAddress, int64(req.Nonce), req.Symbol,
```
Comment (ElliotFriedman) [inline x/ethbridge/client/rest/rest.go:109]: We really shouldn't be having to type cast right here

File: x/ethbridge/client/rest/rest.go [hunk 11]
```diff
@@ -93,33 +96,35 @@ func createClaimHandler(cliCtx context.CLIContext) http.HandlerFunc {
...
 
-		claimType, err := types.StringToClaimType(req.ClaimType)
-		if err != nil {
+		claimType, exist := types.ClaimType_value[req.ClaimType]
+		if !exist {
 			rest.WriteErrorResponse(w, http.StatusBadRequest, types.ErrInvalidClaimType.Error())
 			return
 		}
 
+		ct := types.ClaimType(claimType)
```
Comment (ElliotFriedman) [inline x/ethbridge/client/rest/rest.go:105]: Let's find a better name for this variable

File: x/ethbridge/handler.go [hunk 12]
```diff
@@ -3,315 +3,41 @@ package ethbridge
...
-		log.Fatalln("failed to init zap logging")
-	}
-	return logger.Sugar()
-}
-
 // NewHandler returns a handler for "ethbridge" type messages.
-func NewHandler(
-	accountKeeper types.AccountKeeper, bridgeKeeper Keeper, cdc *codec.Codec) sdk.Handler {
+func NewHandler(k Keeper) sdk.Handler {
+	msgServer := keeper.NewMsgServerImpl(k)
```
Comment (ElliotFriedman) [inline x/ethbridge/handler.go:15]: What keeper is this? I'm assuming it's an ethbridge module keeper.

File: x/ethbridge/keeper/grpc_query.go [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+	prophecy, found := srv.Keeper.oracleKeeper.GetProphecy(sdkCtx, id)
+	if !found {
+		return nil, sdkerrors.Wrap(oracletypes.ErrProphecyNotFound, id)
+	}
+
+	bridgeClaims, err := types.MapOracleClaimsToEthBridgeClaims(
+		req.EthereumChainId,
+		types.NewEthereumAddress(req.BridgeContractAddress),
+		req.Nonce,
```
Comment (ElliotFriedman) [inline x/ethbridge/keeper/grpc_query.go:39]: Any type casting we need to do here before calling the `MapOracleClaimsToEthBridgeClaims` function?

File: x/ethbridge/keeper/grpc_query.go [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+		req.Symbol,
+		types.NewEthereumAddress(req.TokenContractAddress),
+		types.NewEthereumAddress(req.EthereumSender),
+		prophecy.ValidatorClaims,
```
Comment (ElliotFriedman) [inline x/ethbridge/keeper/grpc_query.go:43]: This is incorrect as the cosmos receiver address should be included here when creating the new OracleClaim from EthBridgeClaim

This means we need to modify this function: `CreateEthClaimFromOracleString` to accept a sifchain address.

We need to ensure that the prophecy ID's get constructed exactly the same way they did before to prevent hash collisions.

File: x/ethbridge/keeper/keeper.go [hunk 15]
```diff
@@ -4,36 +4,38 @@ import (
...
 // exposes getter/setter methods for the various parts of the state machine
 type Keeper struct {
-	cdc *codec.Codec // The wire codec for binary encoding/decoding.
+	cdc codec.BinaryMarshaler // The wire codec for binary encoding/decoding.
 
-	supplyKeeper types.SupplyKeeper
-	oracleKeeper types.OracleKeeper
-	storeKey     sdk.StoreKey
+	accountKeeper types.AccountKeeper
+	bankKeeper    types.BankKeeper
```
Comment (ElliotFriedman) [inline x/ethbridge/keeper/keeper.go:26]: Bank keeper is now standing in for supply keeper I take it?

**Review by ElliotFriedman — CHANGES_REQUESTED** [review]

File: x/ethbridge/keeper/msg_server.go [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+	}
+
+	cosmosSender, err := sdk.AccAddressFromBech32(msg.CosmosSender)
+	if err != nil {
+		return nil, err
+	}
+
+	account := srv.Keeper.accountKeeper.GetAccount(ctx, cosmosSender)
+	if account == nil {
+		logger.Error("account is nil.", "CosmosSender", msg.CosmosSender)
```
Comment (ElliotFriedman) [inline x/ethbridge/keeper/msg_server.go:44]: should be `sender account is nil` instead of just `account is nil`.

File: x/ethbridge/keeper/msg_server.go [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeySender, msg.CosmosSender),
+		),
+		sdk.NewEvent(
+			types.EventTypeLock,
+			sdk.NewAttribute(types.AttributeKeyEthereumChainID, strconv.FormatInt(msg.EthereumChainId, 10)),
+			sdk.NewAttribute(types.AttributeKeyCosmosSender, msg.CosmosSender),
```
Comment (ElliotFriedman) [inline x/ethbridge/keeper/msg_server.go:71]: We might need to call the .String() method on the msg.CosmosSender if it has been converted to another type by this point.

File: x/ethbridge/keeper/msg_server.go [hunk 18]
```diff
@@ -0,0 +1,322 @@
...
+			sdk.NewAttribute(types.AttributeKeyCosmosSenderSequence, strconv.FormatUint(account.GetSequence(), 10)),
+			sdk.NewAttribute(types.AttributeKeyEthereumReceiver, msg.EthereumReceiver),
```
Comment (ElliotFriedman) [inline x/ethbridge/keeper/msg_server.go:73]: Same comment here if EthereumReceiver has been converted to an address type, we will need to convert it.

File: x/ethbridge/keeper/msg_server.go [hunk 19]
```diff
@@ -0,0 +1,322 @@
...
+			sdk.NewAttribute(types.AttributeKeyAmount, msg.Amount.String()),
+			sdk.NewAttribute(types.AttributeKeySymbol, msg.Symbol),
+			sdk.NewAttribute(types.AttributeKeyCethAmount, msg.CethAmount.String()),
```
Comment (ElliotFriedman) [inline x/ethbridge/keeper/msg_server.go:76]: Looks like we removed the last item here which was the coins object stringified. It had a key for the event of `AttributeKeyCoins` which had a value of coins.String(). As long as the relayer didn't use this information this is fine.

File: x/ethbridge/keeper/msg_server.go [hunk 20]
```diff
@@ -0,0 +1,322 @@
...
+	})
+
+	return &types.MsgBurnResponse{}, nil
+
+}
+func (srv msgServer) CreateEthBridgeClaim(goCtx context.Context, msg *types.MsgCreateEthBridgeClaim) (*types.MsgCreateEthBridgeClaimResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	logger := srv.Keeper.Logger(ctx)
+	var mutex = &sync.RWMutex{}
```
Comment (ElliotFriedman) [inline x/ethbridge/keeper/msg_server.go:143]: We should remove all of these mutex locks.

File: x/ethbridge/keeper/msg_server.go [hunk 21]
```diff
@@ -0,0 +1,322 @@
...
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeySender, msg.EthBridgeClaim.ValidatorAddress),
+		),
+		sdk.NewEvent(
+			types.EventTypeCreateClaim,
+			sdk.NewAttribute(types.AttributeKeyCosmosSender, msg.EthBridgeClaim.ValidatorAddress),
+			sdk.NewAttribute(types.AttributeKeyEthereumSender, msg.EthBridgeClaim.EthereumSender),
+			sdk.NewAttribute(types.AttributeKeyEthereumSenderNonce, strconv.FormatInt(msg.EthBridgeClaim.Nonce, 10)),
```
Comment (ElliotFriedman) [inline x/ethbridge/keeper/msg_server.go:181]: Let's revert this back to Itoa instead of FormatInt.

File: x/ethbridge/keeper/msg_server.go [hunk 22]
```diff
@@ -0,0 +1,322 @@
...
+		"CethReceiverAccount", msg.CethReceiverAccount)
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeySender, msg.CosmosSender),
+		),
+		sdk.NewEvent(
+			types.EventTypeLock,
```
Comment (ElliotFriedman) [inline x/ethbridge/keeper/msg_server.go:287]: This is an incorrect event name for what this function does.

File: x/ethbridge/keeper/peggyTokens.go [hunk 23]
```diff
@@ -15,14 +15,14 @@ func (k Keeper) AddPeggyToken(ctx sdk.Context, token string) {
 	store := ctx.KVStore(k.storeKey)
 	key := types.PeggyTokenKeyPrefix
 
-	tokens = append(tokens, token)
-	store.Set(key, k.cdc.MustMarshalBinaryBare(tokens))
+	tokens.Tokens = append(tokens.Tokens, token)
```
Comment (ElliotFriedman) [inline x/ethbridge/keeper/peggyTokens.go:18]: Why do we need to make this change to the way we store data in the keeper? We should be able to do it the old way and just append to an array of type tokens without having to use a property of the object to append to.

File: x/ethbridge/keeper/querier_test.go [hunk 24]
```diff
@@ -7,27 +7,25 @@ import (
...
-	cdc := keeperLib.MakeTestCodec()
+	ctx, keeper, _, _, _, encCfg, _ := test.CreateTestKeepers(t, 0.7, []int64{3, 3}, "")
 
 	query := abci.RequestQuery{
 		Path: "",
 		Data: []byte{},
 	}
 
-	querier := NewQuerier(oracleKeeper, cdc)
+	querier := ethbridgekeeper.NewLegacyQuerier(keeper, encCfg.Amino)
```
Comment (ElliotFriedman) [inline x/ethbridge/keeper/querier_test.go:28]: Why are we using amino here? I thought we moved to protobufs

File: x/ethbridge/module.go [hunk 25]
```diff
@@ -34,34 +41,39 @@ func (AppModuleBasic) Name() string {
...
 // DefaultGenesis returns default genesis state as raw bytes for the ethbridge
 // module.
-func (AppModuleBasic) DefaultGenesis() json.RawMessage {
+func (b AppModuleBasic) DefaultGenesis(marshaler codec.JSONMarshaler) json.RawMessage {
 	return nil
 }
 
 // ValidateGenesis performs genesis state validation for the ethbridge module.
-func (AppModuleBasic) ValidateGenesis(_ json.RawMessage) error {
+func (b AppModuleBasic) ValidateGenesis(marshaler codec.JSONMarshaler, config sdkclient.TxEncodingConfig, message json.RawMessage) error {
```
Comment (ElliotFriedman) [inline x/ethbridge/module.go:55]: Why pass all of these parameters if they aren't used?

File: x/ethbridge/module.go [hunk 26]
```diff
@@ -34,34 +41,39 @@ func (AppModuleBasic) Name() string {
...
 	return nil
 }
 
+func (b AppModuleBasic) RegisterGRPCGatewayRoutes(c sdkclient.Context, serveMux *runtime.ServeMux) {
+	// TODO: Register grpc gateway
```
Comment (ElliotFriedman) [inline x/ethbridge/module.go:60]: Is there something else we should be doing here @timlind @Jedi2002?

File: x/ethbridge/module.go [hunk 27]
```diff
@@ -109,36 +126,36 @@ func (am AppModule) RegisterInvariants(ir sdk.InvariantRegistry) {
...
 // QuerierRoute returns the ethbridge module's querier route name.
 func (AppModule) QuerierRoute() string {
 	return QuerierRoute
 }
 
-// NewQuerierHandler returns the ethbridge module sdk.Querier.
-func (am AppModule) NewQuerierHandler() sdk.Querier {
-	return NewQuerier(am.OracleKeeper, am.Codec)
+// Deprecated: LegacyQuerierHandler use RegisterServices
+func (am AppModule) LegacyQuerierHandler(amino *codec.LegacyAmino) sdk.Querier {
```
Comment (ElliotFriedman) [inline x/ethbridge/module.go:144]: Why are we supporting this legacy behavior? I think we should completely just deprecate it if we don't have a use for it.

File: x/ethbridge/types/claim.go [hunk 28]
```diff
@@ -3,43 +3,35 @@ package types
...
+	bridgeContract EthereumAddress,
+	nonce int64,
+	symbol string,
+	tokenContact EthereumAddress,
+	ethereumSender EthereumAddress,
+	cosmosReceiver sdk.AccAddress,
+	validator sdk.ValAddress,
+	amount sdk.Int,
+	claimType ClaimType,
+) *EthBridgeClaim {
```
Comment (ElliotFriedman) [inline x/ethbridge/types/claim.go:25]: Why did we make this function return a pointer to a newly allocated EthBridgeClaim instead of just a new object?

We should roll this back to not using pointers.

File: x/ethbridge/types/claim.go [hunk 29]
```diff
@@ -72,28 +64,39 @@ func NewOracleClaimContent(
...
-	ethereumChainID int, bridgeContract EthereumAddress, nonce int,
-	ethereumAddress EthereumAddress, validator sdk.ValAddress, oracleClaimString string,
-) (EthBridgeClaim, error) {
+	ethereumChainID int64,
+	bridgeContract EthereumAddress,
+	nonce int64,
+	ethereumAddress EthereumAddress,
+	validator sdk.ValAddress,
+	oracleClaimString string,
+) (*EthBridgeClaim, error) {
```
Comment (ElliotFriedman) [inline x/ethbridge/types/claim.go:96]: Same question here: why do we use a pointer to an EthBridgeClaim instead of just a regular EthBridgeClaim.

File: x/ethbridge/types/msgs.go [hunk 30]
```diff
@@ -17,24 +17,14 @@ const (
...
-	Symbol           string          `json:"symbol" yaml:"symbol"`
-	EthereumChainID  int             `json:"ethereum_chain_id" yaml:"ethereum_chain_id"`
-	EthereumReceiver EthereumAddress `json:"ethereum_receiver" yaml:"ethereum_receiver"`
-	CethAmount       sdk.Int         `json:"ceth_amount" yaml:"ceth_amount"`
-}
-
 // NewMsgLock is a constructor function for MsgLock
 func NewMsgLock(
-	ethereumChainID int, cosmosSender sdk.AccAddress,
+	ethereumChainID int64, cosmosSender sdk.AccAddress,
```
Comment (ElliotFriedman) [inline x/ethbridge/types/msgs.go:22]: Should this be a uint64 instead of an int64? I'm having a hard time imagining a chainid or chain descriptor being negative.

File: x/ethbridge/types/msgs.go [hunk 31]
```diff
@@ -49,19 +39,19 @@ func (msg MsgLock) Type() string { return "lock" }
 
 // ValidateBasic runs stateless checks on the message
 func (msg MsgLock) ValidateBasic() error {
-	if strconv.Itoa(msg.EthereumChainID) == "" {
-		return sdkerrors.Wrapf(ErrInvalidEthereumChainID, "%d", msg.EthereumChainID)
+	if strconv.FormatInt(msg.EthereumChainId, 10) == "" {
```
Comment (ElliotFriedman) [inline x/ethbridge/types/msgs.go:42]: Let's switch back to Itoa.

File: cmd/ebrelayer/types/types.go (hunk 6 above)
Comment (ElliotFriedman) [reply cmd/ebrelayer/types/types.go:49]: Non issue. Closing.

File: x/ethbridge/client/cli/tx.go (hunk 9 above)
Comment (ElliotFriedman) [reply x/ethbridge/client/cli/tx.go:54]: Reason for using Parseint instead of Atoi is parseint stores result in larger data container.

File: x/ethbridge/keeper/msg_server.go (hunk 22 above)
Comment (ElliotFriedman) [reply x/ethbridge/keeper/msg_server.go:287]: Peggy team should resolve this as this is legacy code that needs to be fixed.

**Comment by marshsif** [comment]
This PR should probably be closed.

**Comment by timlind** [comment]
This is an old draft PR for 42 work from Marko.
